### ISSUE TITLE
java.lang.Record was not found in Jandex index

### ISSUE THREAD

**Issue by pdolezal** [quarkusio/quarkus#19255]
### Describe the bug

When a Java 16 record class appears in the application being built, Maven produces warnings that `java.lang.Record` is not in Jandex index.

### Expected behavior

There should be no such warnings related to `java.lang.Record`.

### Actual behavior

Maven build produces following warning:

```
[WARNING] [io.quarkus.arc.processor.BeanArchives] Failed to index java.lang.Record: Class does not exist in ClassLoader QuarkusClassLoader:Deployment Class Loader: PROD@22ab76e2
[INFO] [io.quarkus.arc.processor.IndexClassLookupUtils] Class for name: java.lang.Record was not found in Jandex index. Please ensure the class is part of the index.
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 5491ms
```

### Output of `java -version`

16

### Quarkus version or git rev

2.1.0

### Build tool (ie. output of `mvnw --version` or `gradlew --version`)

Maven 3.8.1

**Comment by gsmet** [comment]
Any chance you could provide a small reproducer? Thanks!

**Comment by pdolezal** [comment]
Here is [a small demo application](https://github.com/quarkusio/quarkus/files/6945424/points.zip) which uses a record and Maven build produces the message mentioned in the OP.

I hope it helps.

**Comment by gastaldi** [comment]
@pdolezal I tried to reproduce this using the project you provided but I don't see the warning you mentioned. 

I am using JDK 16.0.2  - the warning also doesn't appear if you change the project to use Quarkus 2.1.1.Final

**Comment by pdolezal** [comment]
I'll be back on the machine on Monday to get the exact version of the JDK. At this moment I can only say it is Oracle JDK 16 on Windows and probably the initial release rather than 16.0.2 update. On the other hand, I'd be surprised if that had any effect. But who knows?

**Comment by gastaldi** [comment]
@pdolezal I built using Windows 10 and 16.0.2 and I still don't see the warning using your sample project

Here is the output:

```bash
C:\Users\george\Downloads\points>mvn -V clean package
Apache Maven 3.6.2 (40f52333136460af0dc0d7232c0dc0bcf0d9e117; 2019-08-27T12:06:16-03:00)
Maven home: C:\apache-maven-3.6.2\bin\..
Java version: 16.0.2, vendor: Oracle Corporation, runtime: C:\Users\george\Downloads\jdk-16.0.2
Default locale: pt_BR, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
[INFO] Scanning for projects...
[INFO]
[INFO] -------------------------< com.example:points >-------------------------
[INFO] Building points 1.0.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ points ---
[INFO] Deleting C:\Users\george\Downloads\points\target
[INFO]
[INFO] --- quarkus-maven-plugin:2.1.1.Final:generate-code (default) @ points ---
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ points ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 1 resource
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ points ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 1 source file to C:\Users\george\Downloads\points\target\classes
[INFO]
[INFO] --- quarkus-maven-plugin:2.1.1.Final:generate-code-tests (default) @ points ---
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ points ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory C:\Users\george\Downloads\points\src\test\resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ points ---
[INFO] No sources to compile
[INFO]
[INFO] --- maven-surefire-plugin:2.12.4:test (default-test) @ points ---
[INFO] No tests to run.
[INFO]
[INFO] --- maven-jar-plugin:3.2.0:jar (default-jar) @ points ---
[INFO] Building jar: C:\Users\george\Downloads\points\target\points-1.0.0-SNAPSHOT.jar
[INFO]
[INFO] --- quarkus-maven-plugin:2.1.1.Final:build (default) @ points ---
[INFO] [org.jboss.threads] JBoss Threads version 3.4.0.Final
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 9427ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  28.823 s
[INFO] Finished at: 2021-08-06T15:15:15-03:00
[INFO] ------------------------------------------------------------------------

```

**Comment by pdolezal** [comment]
I investigated the issue for a while and I came to something.

I use JDK 11 as my default JDK, while I have multiple other JDK installations in parallel. I use `maven-toolchains-plugin`, so that I can run builds for non-default Java versions. The POM refers to the plugin, my `toolchains.xml` refer then to all my JDK versions and `settings.xml` auto-activate `toolchains` profile that the POM contains as well. Therefore Maven picks the right JDK toolchain, no matter under which Java version it was launched.

The warning, which I reported, appears in my usual situation when Maven is launched with JDK 11, but toolchains pick JDK 16 for the actual build. When I change my `JAVA_HOME` to point to the same JDK as toolchains should choose, the warning does not appear.

Hence I conclude that Quarkus plugin ignores toolchains.

**Comment by gastaldi** [comment]
> Hence I conclude that Quarkus plugin ignores toolchains.

@pdolezal that's an interesting find. @aloubyansky can you confirm that?

**Comment by aloubyansky** [comment]
It should work in dev mode but not when packaging. There is https://github.com/quarkusio/quarkus/issues/12943 I am actually not looking into that at the moment. I originally assigned it to myself to fix the dev mode regression.

**Comment by gastaldi** [comment]
Closing this as a duplicate of https://github.com/quarkusio/quarkus/issues/12943
